### PR TITLE
Installation script

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR concerns the first implementation of the installation process related to the issue #2299 

- [ ] tests : 
  - [ ] end-to-end
  - [x] integration

### CODE REVIEW

File: client/src/modules/install/install.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+// controller definition
+function InstallApplicationController(InstallService, $state, Notify) {
+  var vm = this;
+
+  vm.setup = { enterprise : {}, project : {}, user : {} };
+
+  // expose methods to the view
+  vm.submit = submit;
+
+  InstallService.checkBasicInstallExist()
```
Comment (jniles) [inline client/src/modules/install/install.js]: This is pretty cool!

It looks like what this code will do is basically this:
navigate to /install -> load InstallApplicationController -> fire HTTP GET /install -> if already exist -> $state.go('login').

You could probably remove the step that loads the install controller if the `alreadyExist` flag is false in `onEnter` part of the `$state` change.  Something like this:
```js
onEnter: ['InstallService', '$state', function (InstallService, $state) {
  return InstallService.checkBasicInstallExist()
     .then(function (isInstalled) { 
       if (isInstalled) { $state.go('login'); }
     });
}]
```

Trying to catch the `isInstalled` flag before navigating to the route should make sure that there isn't a flash of content before re-routing.

File: client/src/modules/install/install.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  function notifyInstallSucess() {
+    Notify.success('INSTALL.SUCCESS_INSTALL');
+    $state.go('login');
+  }
+
+  function checkPassword() {
+    if (!vm.setup.user.password) { return false; }
+
+    if (vm.setup.user.password !== vm.setup.user.repassword) {
```
Comment (jniles) [inline client/src/modules/install/install.js]: Do we not have a service for this?  Or a directive/component?

File: client/src/modules/install/install.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    return true;
+  }
+
+  function submit(form) {
+    if (form.$invalid) { return 0; }
+
+    if (checkPassword() === false) { return Notify.warn('USERS.PASSWORD_MISMATCH'); }
+
+    return InstallService.proceedInstall(vm.setup)
+      .then(notifyInstallSucess)
```
Comment (jniles) [inline client/src/modules/install/install.js]: Nice function names 👍

File: server/app.js [hunk 4]
```diff
@@ -25,11 +25,16 @@
 // Workaround - required before 'use-strict'
 const json2xls = require('json2xls');
 
+
 require('use-strict');
+const dotEnv = require('dotenv');
```
Comment (jniles) [inline server/app.js:30]: :+1:

File: server/controllers/install.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+ */
+function basicInstallExist() {
+  // check users, enterprise, projects
+  const userExist = 'SELECT COUNT(*) AS total_users FROM user;';
+  const enterpriseExist = 'SELECT COUNT(*) AS total_enterprises FROM enterprise;';
+  const projectExist = 'SELECT COUNT(*) AS total_projects FROM project;';
+
+  const dbPromise = [db.one(userExist), db.one(enterpriseExist), db.one(projectExist)];
+
+  return Q.all(dbPromise)
```
Comment (jniles) [inline server/controllers/install.js]: Nice!

File: server/controllers/install.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ * @description
+ * check if the basic information for installation exist, in the case
+ * it doesn't exist we can perform a new installation
+ */
+function basicInstallExist() {
+  // check users, enterprise, projects
+  const userExist = 'SELECT COUNT(*) AS total_users FROM user;';
+  const enterpriseExist = 'SELECT COUNT(*) AS total_enterprises FROM enterprise;';
+  const projectExist = 'SELECT COUNT(*) AS total_projects FROM project;';
```
Comment (jniles) [inline server/controllers/install.js]: I know JavaScript will treat 1 as `true`, but why don't you just go ahead and make the SQL return a boolean?  Like:
```js
const userExist = 'SELECT COUNT(*) > 1 AS hasUsers FROM user;'
```
or something?

File: server/controllers/install.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * POST /install
+ *
+ * proceed to the application installation
+ */
+exports.proceedInstall = (req, res, next) => {
+  const { enterprise, project, user } = req.body;
+
+  basicInstallExist()
+    .then(ans => {
```
Comment (jniles) [inline server/controllers/install.js]: Instead of `ans` a better name would be `isInstalled`.

File: server/controllers/install.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    // prepare transactions
+    const transaction = db.transaction();
+    const sqlEnterprise = 'INSERT INTO enterprise SET ? ';
+    const sqlProject = 'INSERT INTO project SET ? ';
+    const sqlUser = `
+      INSERT INTO user (username, password, display_name) VALUES (?, PASSWORD(?), ?);`;
+
+    const sqlUnitPermission = `
```
Comment (jniles) [inline server/controllers/install.js:108]: :+1:

**Review by jniles — COMMENTED** [review]
Awesome work!  The code is very readable, you have nice variable names, nice refactoring to help ESLint along.  Overall, really great work.

I've made a few comments to try and improve the logic and get rid of that flash of unrendered content.  Let me know if it works or not.  We might also explore a method using `resolve` instead of `onEnter` to resolve a promise before navigating and change the behavior based on that.

It doesn't look like `test/data-init.sql` is used ... if you are going to use it for integration tests or end to end tests, that is good.

**Comment by mbayopanda** [comment]
@jniles could you review this PR ?

File: client/src/modules/install/install.js (hunk 2 above)
Comment (jniles) [reply client/src/modules/install/install.js]: It looks like [UserService.validatePassword()](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/modules/users/users.service.js#L124) will do this for you.  Could you use that?

File: test/end-to-end/locations/provinces.spec.js [hunk 9]
```diff
@@ -1,25 +1,26 @@
...
+describe('Provinces Management', () => {
 
   const path = '#!/locations/province';
 
   before(() => helpers.navigate(path));
 
   const province = {
-    country : 'Test Hook Country',
-    name : 'New Province'
+    country : 'République Démocratique du Congo',
```
Comment (jniles) [inline test/end-to-end/locations/provinces.spec.js:19]: Nice cleanup :+1:

File: test/integration/locations.js [hunk 10]
```diff
@@ -262,8 +250,8 @@ describe('(/locations) Locations Interface', function () {
   });
 
   it('PUT /locations/countries/:uuid should update a Country', function () {
-    return agent.put('/locations/countries/' + country.uuid)
-      .send({ name : 'Update New Country'})
+    return agent.put(`/locations/countries/${country.uuid}`)
```
Comment (jniles) [inline test/integration/locations.js:253]: Nice cleanup! 👍

**Review by jniles — APPROVED** [review]
Code looks pretty good!  I've made a few non-blocking comments.

How do I test this locally?

**Comment by jniles** [comment]
Looks like everything is addressed and ready to work here.

bors r+

**Pull request merged**